### PR TITLE
Implemented audio input support for JavaScript audio driver

### PR DESCRIPTION
**Pull request by marcelofg55**
Note, I've added the flag `BINARYEN_TRAP_MODE='clamp'` since I was getting "float unrepresentable in integer range" error without it.

### CODE REVIEW

File: platform/javascript/audio_driver_javascript.cpp [hunk 1]
```diff
@@ -152,14 +175,64 @@ void AudioDriverJavaScript::finish() {
...
+	/* clang-format on */
+
+	return OK;
+}
+
+Error AudioDriverJavaScript::capture_stop() {
+
+	/* clang-format off */
+	EM_ASM({
+		_audioDriver_audioInput.disconnect();
```
Comment (eska014) [inline platform/javascript/audio_driver_javascript.cpp]: We should `stop()` the MediaStreamTrack here to release the mic

File: platform/javascript/audio_driver_javascript.cpp [hunk 2]
```diff
@@ -152,14 +175,64 @@ void AudioDriverJavaScript::finish() {
...
+
+	if (internal_buffer) {
+		memdelete_arr(internal_buffer);
+		internal_buffer = NULL;
+	}
+}
+
+Error AudioDriverJavaScript::capture_start() {
+
+	input_buffer.resize(buffer_length * 2 * 8);
```
Comment (eska014) [inline platform/javascript/audio_driver_javascript.cpp]: Why multiply with 8?

**Review by leonkrause — CHANGES_REQUESTED** [review]

File: platform/javascript/audio_driver_javascript.cpp [hunk 3]
```diff
@@ -101,20 +114,30 @@ void AudioDriverJavaScript::start() {
 		var INTERNAL_BUFFER_PTR = $0;
 
 		var audioDriverMixFunction = cwrap('audio_driver_js_mix');
+		var audioDriverProcessCapture = cwrap('audio_driver_process_capture');
```
[the remark this answers is not in the document]
Comment (marcelofg55) [reply platform/javascript/audio_driver_javascript.cpp]: Fixed.

File: platform/javascript/audio_driver_javascript.cpp [hunk 4]
```diff
@@ -44,38 +44,50 @@ extern "C" EMSCRIPTEN_KEEPALIVE void audio_driver_js_mix() {
...
 	audio_server_process(sample_count, stream_buffer);
 	for (int i = 0; i < sample_count * channel_count; i++) {
-		internal_buffer[i] = float(stream_buffer[i] >> 16) / 32768.0;
+		internal_buffer[i] = float(stream_buffer[i] >> 16) / 32768.f;
 	}
 }
 
+void AudioDriverJavaScript::process_capture(float sample) {
+
+	int32_t sample32 = int32_t(sample * 32768.f) << 16;
```
[the remark this answers is not in the document]
Comment (marcelofg55) [reply platform/javascript/audio_driver_javascript.cpp]: Fixed.

File: platform/javascript/audio_driver_javascript.cpp (hunk 2 above)
Comment (marcelofg55) [reply platform/javascript/audio_driver_javascript.cpp]: I've changed this now but the reason was that input_buffer works as a ring buffer so it needs to be larger.

File: platform/javascript/audio_driver_javascript.cpp (hunk 1 above)
Comment (marcelofg55) [reply platform/javascript/audio_driver_javascript.cpp]: Added.

File: platform/javascript/detect.py [hunk 5]
```diff
@@ -122,6 +122,7 @@ def configure(env):
     ## Link flags
 
     env.Append(LINKFLAGS=['-s', 'BINARYEN=1'])
+    env.Append(LINKFLAGS=['-s', 'BINARYEN_TRAP_MODE=\'clamp\''])
```
[the remark this answers is not in the document]
Comment (marcelofg55) [reply platform/javascript/detect.py]: I haven't split a commit before with git so if it's not that important I will leave it as it is.

File: platform/javascript/audio_driver_javascript.cpp (hunk 1 above)
Comment (marcelofg55) [reply platform/javascript/audio_driver_javascript.cpp]: Just fixed it.

**Review by leonkrause — APPROVED** [review]

**Comment by akien-mga** [comment]
Now that Godot 3.1 has been released, we don't plan to cherry-pick new features and enhancements to the 3.0 branch, unless there is very strong support in favor of it. Removing cherrypick label for 3.0.

**Pull request merged**